File: 0061-rotate-list/0061-rotate-list.java
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) return head;

        // Step 1: Calculate the length of the list
        ListNode curr = head;
        int length = 1; // at least 1 node
        while (curr.next != null) {
            curr = curr.next;
            length++;
        }

        // Step 2: Connect the tail to the head to make it circular
        curr.next = head;

        // Step 3: Normalize k
        k = k % length;
        if (k == 0) {
            curr.next = null; // break the circle
            return head;
        }

        // Step 4: Find the new tail = length - k - 1 steps from head
        ListNode newTail = head;
        for (int i = 0; i < length - k - 1; i++) {
            newTail = newTail.next;
        }

        // Step 5: Set the new head and break the circle
        ListNode newHead = newTail.next;
        newTail.next = null;

        return newHead;
    }
}

File: 0061-rotate-list/0061-rotate-list.cpp
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == NULL || head->next == NULL || k == 0)
            return head;

        // Step 1: find length
        int len = 1;
        ListNode* tail = head;
        while(tail->next != NULL){
            tail = tail->next;
            len++;
        }

        // Step 2: reduce k
        k = k % len;
        if(k == 0) return head;

        // Step 3: make circular
        tail->next = head;

        // Step 4: find new tail
        int steps = len - k;
        ListNode* newTail = head;
        for(int i = 1; i < steps; i++){
            newTail = newTail->next;
        }

        // Step 5: break circle
        ListNode* newHead = newTail->next;
        newTail->next = NULL;

        return newHead;
    }
};
